flow/layers/container_layer.cc

```diff
@@ -49,6 +49,11 @@ void ContainerLayer::PaintChildren(PaintContext& context) const {
   }
 }
 
+float ContainerLayer::get_total_elevation() const {
+  return elevation_ +
+         (parent() != nullptr ? parent()->get_total_elevation() : 0.0f);
```

Comment:
Shall we cache this `total_elevation` inside the `ContainerLayer` so we can spend O(1) time querying it instead of O(n) time (walk up all the way to the root) for each query? For example, if all elevations are set before preroll, we can use O(n) time to compute all the total elevations during preroll and just cache the results.

It would be Ok to leave this O(n) implementation as is if this is rarely called. But make sure to add a clear comment to the function signature that this is O(n) instead of O(1) because most people would probably expect O(1) for a `get_something` method.

Reply:
I don't think we safely could without a larger refactor on how these are used. If somebody changes the elevation of an ancestor of this layer, the cached value would be invalid. We don't have any hooks for when the parent's elevation is updated here, and from the parent side we don't have the ability to walk through children and update them once its elevation changes. I could say not to change the elevation of any layer once it has children in the documentation, but that seems brittle.I could add in callback handlers for when the value updates here, but I'm not totally sure if that would be worth it. WDYT?

Reply:
Is `GetTotalElevation` currently only used inside `UpdateScene` which is guaranteed to be called after `Preroll`? If so, we could update total elevations for all nodes in `Preroll` and use them safely in `UpdateScene`. This is similar to `paint_bounds` which is computed in `Preroll` and it's only safe to use after `Preroll`. 

According to the code change in this PR, the O(n) complexity could be an issue for a layer tree that has a long chain of n `PhysicalShapeLayer`s. It would require O(n^2) time in that case to walk all the `PhysicalShapeLayer` in `UpdateScene`. I think that we try to avoid any quadratic complexity in all our tree walks (e.g., single-pass layout instead of a multi-pass layout).

Reply:
The caching does work fine how the code is currently written. I was wary of caching this in a way that would silently break if the call order were to change. I get that the performance problem is a bigger concern though. If this is already happening in other places in the code it's less of a usability issue too, I didn't realize some of these were already only safe to use at certain times. Updated.